<launch>
    <!-- Arguments -->
    <arg name="cmd_vel_topic" default="/cmd_vel" />
    <arg name="odom_topic" default="odom" />
    <arg name="move_forward_only" default="false"/>
    <arg name="map_file" default="$(find snow_removal)/maps/map.yaml"/>

    <arg name="offline"          default="false"/>
    <include unless="$(arg offline)" 
        file="$(find realsense2_camera)/launch/rs_camera.launch">
        <arg name="align_depth" value="true"/>
        <arg name="linear_accel_cov" value="1.0"/>
        <arg name="unite_imu_method" value="linear_interpolation"/>
    </include>
    

    <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>
    
    <node pkg="imu_filter_madgwick" type="imu_filter_node" name="ImuFilter">
        <param name="use_mag" type="bool" value="false" />
        <param name="_publish_tf" type="bool" value="false" />
        <param name="_world_frame" type="string" value="enu" />
        <remap from="/imu/data_raw" to="/camera/imu"/>
    </node>

    

    <include file="$(find rtabmap_ros)/launch/rtabmap.launch">
	    <arg name="args" default="--Vis/MinInliers 15 --Mem/IncrementalMemory false --Mem/InitWMWithAllNodes true"/>
	    <arg name="localization" value="true"/>
        <arg name="rgb_topic" value="/camera/color/image_raw"/>
        <arg name="depth_topic" value="/camera/aligned_depth_to_color/image_raw"/>
        <arg name="camera_info_topic" value="/camera/color/camera_info"/>
        <arg name="depth_camera_info_topic" value="/camera/depth/camera_info"/>
        <arg name="rtabmapviz" value="false"/>
        <arg name="rviz" value="true"/>
	</include>

    <include file="$(find robot_localization)/launch/ukf_template.launch"/>
    <param name="/ukf_se/frequency" value="300"/>
    <param name="/ukf_se/base_link_frame" value="camera_link"/>
    <param name="/ukf_se/odom0" value="rtabmap/odom"/>
    <rosparam param="/ukf_se/odom0_config">[true,true,true,
                                            true,true,true,
                                            true,true,true,
                                            true,true,true,
                                            true,true,true]
    </rosparam>
    <param name="/ukf_se/odom0_relative" value="true"/>
    <param name="/ukf_se/odom0_pose_rejection_threshold" value="10000000"/>
    <param name="/ukf_se/odom0_twist_rejection_threshold" value="10000000"/>

    <param name="/ukf_se/imu0" value="/imu/data"/>
    <rosparam param="/ukf_se/imu0_config">[false, false, false,
                                           true,  true,  true,
                                           true,  true,  true,
                                           true,  true,  true,
                                           true,  true,  true]
    </rosparam>
    <param name="/ukf_se/imu0_differential" value="true"/>
    <param name="/ukf_se/imu0_relative" value="false"/>
    <param name="/ukf_se/use_control" value="false"/>
    <!-- <param name="/ukf_se/odom0_config" value="{true,true,true,}"/> -->

    
    <!-- move_base -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
        <param name="odom_topic" value="odom"/>
        <param name="pose_topic" value="/rtabmap/localization_pose"/>
        <param name="map_topic" value="/rtabmap/octomap_grid"/>
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
        <rosparam file="$(find snow_removal)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find snow_removal)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find snow_removal)/param/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find snow_removal)/param/global_costmap_params.yaml" command="load" />
        <rosparam file="$(find snow_removal)/param/move_base_params.yaml" command="load" />
        <rosparam file="$(find snow_removal)/param/dwa_local_planner_params.yaml" command="load" />
        <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
        <remap from="odom" to="$(arg odom_topic)"/>
        <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />
    </node>

    <!--node pkg="snow_removal" type="autonomous_clearing.py" name="autonomous_clearing"/-->

</launch>